ruby/neo4j/driver/internal/async/pool/connection_pool_impl.rb

```diff
@@ -31,7 +31,7 @@ def acquire(address)
           def retain_all(addresses_to_retain)
             @address_to_pool_lock.with_write_lock do
               @address_to_pool.each do |address, pool|
-                unless addresses_to_retain.include?(address)
+                unless addresses_to_retain.flatten.include?(address)
```

Comment:
Assuming ...
1. `addresses_to_retain` is a hash.
2. `address` is the key in each iteration.
3. `address` and `pool` have different data types.

This may be a premature optimization, but it's easy to make. Do

```ruby
addresses_to_retain.keys.include?(address)
```

It'll make smaller arrays, using less memory and be slightly faster.

Both calls make a new array each time, though. Assuming this method is thread-safe, the value is not changing inside that loop. Regardless of whether you use `#flatten` or `#keys`, you might also consider - for the same reason - 

```ruby
def retain_all(addresses_to_retain)
  addresses = addresses_to_retain.keys
  
  @address_to_pool_lock.with_write_lock do
    @address_to_pool.each do |address, pool|
      unless addresses.include?(address)
```

That way, you create an array of addresses just once.

Reply:
`addresses_to_retain` is a `Set<BoltServerAddress>` which, due to some upstream mechanism, also adds `Set<BoltServerAddres>` to the set which makes `Set<Set<BoltServerAddress>, BoltServerAddress>` which causes the issue.

We can't access the `keys` here since it was similar to how we got to this issue in the first place. The `include?` method was taking the `Set` and trying to access  `attributes` as it would on a `BoltServerAddress` since there were nested sets. 

There might be improvements we can make to how the comparison is done here to make it more performant, but the `Set` class might be doing some heavy lifting for us under the hood

Reply:
Gross!

But thanks for the detailed explanation